### PR TITLE
Move new_data_part shared pointer to allow Outdated part to be deleted immediately on mutation failure

### PR DESCRIPTION
**Pull request by nvartolomei**
### Changelog category (leave one):
- Bug Fix

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix mutations not making progress when checksums do not match between replicas (e.g. caused by a change in data format on an upgrade).

---

This may happen when upgrading from a version before v21.8 (before  #27250) and running a concurrent mutations.

PR without test: #36897

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

File: src/Storages/MergeTree/MergeTreeSettings.h [hunk 1]
```diff
@@ -137,6 +137,9 @@ struct Settings;
     M(UInt64, min_bytes_to_rebalance_partition_over_jbod, 0, "Minimal amount of bytes to enable part rebalance over JBOD array (0 - disabled).", 0) \
     M(Bool, check_sample_column_is_correct, true, "Check columns or columns by hash for sampling are unsigned integer.", 0) \
     \
+    /** Flags that are relevant only in testing. */ \
+    M(Bool, testing_mutate_corrupt_checksums, false, "Tetsing only! Corrupt part checksums during mutation to test fetch fallback.", 0) \
```
Comment (nikitamikhaylov) [inline src/Storages/MergeTree/MergeTreeSettings.h:141]: Maybe better to corrupt checksums by ourselves ?

**Comment by alexey-milovidov** [comment]
@nvartolomei maybe it is possible to throw an exception using throwing functions inside mutation?

**Comment by nvartolomei** [comment]
@alexey-milovidov should be much better now if it passes tests.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Also, add a comment about `hostname` in the test, please.

**Comment by nvartolomei** [comment]
I realized that the test could be rewritten with starting a single node for better efficiency/speed and detach/attach replica table instead of stopping/starting a node. If not even as a stateless test.

But I don't have too much spare time now to do that so I'll leave it as is.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@nvartolomei the integration test did not pass, need to check why.

**Comment by nvartolomei** [comment]
@alexey-milovidov can't find the logs, can you share the link please?

**Comment by alexey-milovidov** [comment]
The change is good to me, but something is wrong with the integration tests.

**Comment by alexey-milovidov** [comment]
@tavplubix check this: https://github.com/ClickHouse/ClickHouse/pull/44059/files#diff-e79ccea43c1f4dfc9eb10b50f8ab8fb81daddac8d1923a06c4ea618c6e44f6a8R372

**Pull request merged**